core/mutator/entry/entry.go

```diff
@@ -41,25 +42,41 @@ func New() *Entry {
 
 // Mutate verifies that this is a valid mutation for this item and applies
 // mutation to value.
-func (*Entry) Mutate(oldValue, mutation []byte) ([]byte, error) {
-	update := new(tpb.SignedKV)
-	if err := proto.Unmarshal(mutation, update); err != nil {
-		return nil, err
-	}
-
+func (*Entry) Mutate(oldValueM, updateM proto.Message) ([]byte, error) {
 	// Ensure that the mutation size is within bounds.
-	if proto.Size(update) > mutator.MaxMutationSize {
-		glog.Warningf("mutation (%v bytes) is larger than the maximum accepted size (%v bytes).", proto.Size(update), mutator.MaxMutationSize)
+	if proto.Size(updateM) > mutator.MaxMutationSize {
+		glog.Warningf("mutation (%v bytes) is larger than the maximum accepted size (%v bytes).", proto.Size(updateM), mutator.MaxMutationSize)
 		return nil, mutator.ErrSize
 	}
 
+	update, ok := updateM.(*tpb.SignedKV)
+	if !ok {
+		glog.Warning("received proto.Message is not of type *tpb.SignedKV.")
+		return nil, fmt.Errorf("updateM.(*tpb.SignedKV): _, %v", ok)
+	}
+	var oldEntry *tpb.Entry
+	if oldValueM != nil {
+		oldEntry, ok = oldValueM.(*tpb.Entry)
+		if !ok {
+			glog.Warning("received proto.Message is not of type *tpb.Entry.")
+			return nil, fmt.Errorf("oldValueM.(*tpb.Entry): _, %v", ok)
+		}
+	}
+
 	// Verify pointer to previous data.
 	// The very first entry will have oldValue=nil, so its hash is the
 	// ObjectHash value of nil.
-	prevEntryHash := objecthash.ObjectHash(oldValue)
+	prevEntryHash := objecthash.ObjectHash(oldEntry)
 	if !bytes.Equal(prevEntryHash[:], update.Previous) {
+		var entryData []byte
+		if oldEntry != nil {
+			var err error
+			if entryData, err = proto.Marshal(oldEntry); err != nil {
+				return nil, fmt.Errorf("Marshal(%v)=%v", oldEntry, err)
+			}
+		}
 		// Check if this mutation is a replay.
-		if bytes.Equal(oldValue, update.GetKeyValue().Value) {
+		if bytes.Equal(entryData, update.GetKeyValue().Value) {
```

Comment:
`proto.Marshal` is not required to produce consistent byte output.  Proto's binary encoding for example, could reorder tags and values.  Could this be upgraded to a `proto.Equal`?